### PR TITLE
Add a client credential flow implementation to the WebAPP

### PR DESCRIPTION
**Pull request by zhichengliu12581**
Below are some of the relevant changes or additions that PR has made:

- The implementation of adding client credential flow is in the Web APP. 
- Add the UT for Client Credential Configuration.

related pr:https://github.com/Azure/azure-sdk-for-java/pull/19808

### CODE REVIEW

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADWebAppConfiguration.java [hunk 1]
```diff
@@ -147,9 +150,13 @@ public static int resourceServerCount(Set<String> scopes) {
...
     private ClientRegistration createClientBuilder(String id, AuthorizationClientProperties authz) {
-        ClientRegistration.Builder result = createClientBuilder(id);
+        ClientRegistration.Builder result = createClientBuilder(id, authz.getAuthorizationGrantType());
         List<String> scopes = authz.getScopes();
-        if (authz.isOnDemand()) {
+        if (AADAuthorizationGrantType.ON_BEHALF_OF.equals(authz.getAuthorizationGrantType())) {
+            throw new IllegalStateException("Web Application do not support on behalf of");
+        }
+        if (authz.isOnDemand()
+            && !AADAuthorizationGrantType.ON_BEHALF_OF.equals(authz.getAuthorizationGrantType())) {
```
[the remark this answers is not in the document]
Comment (chenrujun) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADWebAppConfiguration.java:159]: There are some other invalid configurations:
1. For web-app, `authorizationGrantType = on-behalf-of`.
2. For resource-server, `authorizationGrantType = authorization_code`.

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java [hunk 2]
```diff
@@ -358,6 +360,17 @@ public void afterPropertiesSet() throws Exception {
...
+        authorizationClients.values()
+                            .stream()
+                            .filter(AuthorizationClientProperties::isOnDemand)
+                            .filter(v -> Objects.nonNull(v.getAuthorizationGrantType()))
+                            .forEach(v -> {
+                                if (!AADAuthorizationGrantType.AUTHORIZATION_CODE.equals(
+                                    v.getAuthorizationGrantType())) {
+                                    throw new IllegalStateException("onDemand only support authorization_code grant type");
+                                }
+                            });
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java:373]:
```
authorizationClients
    .values()
    .stream()
    .filter(AuthorizationClientProperties::isOnDemand)
    .map(AuthorizationClientProperties::getAuthorizationGrantType)
    .filter(Objects::nonNull)
    .filter(type -> !AADAuthorizationGrantType.AUTHORIZATION_CODE.equals(type))
    .findAny()
    .ifPresent(notUsed -> {
        throw new IllegalStateException("onDemand only support authorization_code grant type. ");
    });
```

File: sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapp/AADWebAppConfigurationTest.java [hunk 3]
```diff
@@ -133,6 +134,52 @@ public void clientRequiresOnDemandPermissions() {
...
+            });
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void clientWhichIsNotAuthorizationCodeButOnDemandExceptionTest() {
+        WebApplicationContextRunnerUtils
+            .getContextRunnerWithRequiredProperties()
+            .withPropertyValues(
+                "azure.activedirectory.authorization-clients.graph.scopes = fakeValue:/.default",
+                "azure.activedirectory.authorization-clients.graph.authorizationGrantType = on-behalf-of",
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapp/AADWebAppConfigurationTest.java:175]: 1. Change `on-behalf-of` to `client_credentials`.
2. Add similar test in AADResourceServerClientConfigurationTest.

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADClientRegistrationRepository.java [hunk 4]
```diff
@@ -68,6 +71,14 @@ public static boolean isDefaultClient(ClientRegistration clientRegistration) {
             clientRegistration.getClientName());
     }
 
+    public boolean isClientCredentials(String clientId) {
```
Comment (backwind1233) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADClientRegistrationRepository.java:74]: clientId? I think it should be registrationId if it call `allClients.get()`

**Review by backwind1233 — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADClientRegistrationRepository.java (hunk 4 above)
Comment (lzc-1997-abel) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADClientRegistrationRepository.java]: I have fixed it just now

**Review by chenrujun — APPROVED** [review]
LGTM.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapi/AADResourceServerClientConfiguration.java [hunk 5]
```diff
@@ -79,10 +79,20 @@ public OAuth2AuthorizedClientRepository oAuth2AuthorizedClientRepository(
...
+     * Create clients based on configuration items
+     *
+     * @throws IllegalStateException throw if AuthorizationGrantType is authorization_code
+     * @return result of created Clients
+     */
     public List<ClientRegistration> createClients() {
         List<ClientRegistration> result = new ArrayList<>();
         for (String id : properties.getAuthorizationClients().keySet()) {
             AuthorizationClientProperties authorizationProperties = properties.getAuthorizationClients().get(id);
+            if (AADAuthorizationGrantType.AUTHORIZATION_CODE.equals(authorizationProperties.getAuthorizationGrantType())) {
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapi/AADResourceServerClientConfiguration.java:92]: @chenrujun we should further discuss this.

**Pull request merged**